wallet/src/Pos/Util/BackupPhrase.hs

```diff
@@ -29,17 +29,36 @@ newtype BackupPhrase = BackupPhrase
     { bpToList :: [Text]
     } deriving (Eq, Generic)
 
+data MnemonicWordCount
+    = Nine
+    | Twelve
+    | Fifteen
+    | Eighteen
+    | Twentyone
+    | Twentyfour
+    deriving (Eq, Show)
+
+byteCount :: MnemonicWordCount -> Int
+byteCount wc = case wc of
+    Nine -> 3
+    Twelve ->  4
+    Fifteen ->  5
+    Eighteen ->  6
+    Twentyone -> 7
+    Twentyfour -> 8
+
 instance Arbitrary BackupPhrase where
     arbitrary = do
-        em <- arbitraryMnemonic 16
+        em <- arbitraryMnemonic Twelve
         case em of
             Left _  -> arbitrary
             Right a -> pure a
     shrink    = genericShrink
 
-arbitraryMnemonic :: Int -> Gen (Either Text BackupPhrase)
-arbitraryMnemonic len = do
-    eitherMnemonic <- toMnemonic . BS.pack <$> vectorOf len arbitrary
+-- | Generate an arbitrary mnemonic with the given number of words.
+arbitraryMnemonic :: MnemonicWordCount -> Gen (Either Text BackupPhrase)
+arbitraryMnemonic wordCount = do
+    eitherMnemonic <- toMnemonic . BS.pack <$> vectorOf (byteCount wordCount) arbitrary
```

Comment:
I might be undercaffeinated but this is not quite right, isn't it? You are saying that we want a vector of `byteCount Twelve`, which is 4, but note how `byteCount` gives you back the size of the _checksum_, which are _extra bytes_ which needs to be added to the total size.

Therefore, what you really want is `vectorOf ((byteCount wordCount) + 12))`.

I suggest we write also a function that given a `MnemonicWordCount` gives you back the size in bytes. If you take a look at that BIP39 table:

```
| ENT | CS | ENT+CS | MS |
+-------+----+--------+------+
| 128 | 4 | 132 | 12 |
| 160 | 5 | 165 | 15 |
| 192 | 6 | 198 | 18 |
| 224 | 7 | 231 | 21 |
| 256 | 8 | 264 | 24 |
```

Here `byteCount` gives you back only `CS`, but what we need to feed into `vectorOf` is `CS + MS` 😉 

If you take a look at the original implementation, that was exactly where this 16 was coming from for a 12-word mnemonic: 12 `Word8` octets + 4 bytes for the checksum .

Reply:
Thanks for catching that. I will reproduce the table as a function :)